rfcs/20230622-quantized-reduction.md

```diff
@@ -0,0 +1,521 @@
+* (C?) `0 < size(inputs) = size(init_values) = size(results) = N`.
+* (C?) `0 <= dimensions < rank(inputs[0])`.
+* (C?) `is_unique(dimensions)`.
+* (C?) `input_conversion` has type `tensor<E0>, ..., tensor<EN-1> ->
+       (tensor<E'0>, ..., tensor<E'N-1>)` where `Ei = element_type(inputs[i])`.
+* (C?) `body` has type `tensor<E0>, ..., tensor<EN-1>, tensor<E0>, ...,`
+       `tensor<EN-1>) -> (tensor<E0>, ..., tensor<EN-1>)` where
+       `Ei = element_type(output_types(input_conversion)[i])`.
+* (C?) `output_conversion` has type `tensor<E0>, ..., tensor<EN-1> ->
+       (tensor<E'0>, ..., tensor<E'N-1>)` where
+       `E'i = element_type(results[i])`.
+* (C?) `element_type(output_types(input_conversion)...) =
+       element_type(input_types(output_conversion)...)`.
+* (C?) `shape(results...) = shape(inputs...)` except that the dimension
+  sizes of `inputs...` corresponding to `dimensions` are not included.
+```
+
+The above specification of `reduce` op can be used to define the specification
+of other ops as shown below. For brevity, we are only presenting the relevant
+portions of the spec which needs modification.
+
+### reduce_window
+
+```python
+#### Semantics
+
+Applies a reduction functions `input_conversion`, `body`, and
+`output_conversion` to windows of `inputs` and `init_values` and produces
+`results`.
+
+...
+
+More formally,
+`results...[result_index] = reduce(windows, init_values, axes(inputs...),
+        input_conversion, body, output_conversion)`
+where:
+....
+
+#### Inputs
+
+| Label | Name                | Type     |
+|-------|---------------------|----------|
+| (I?)  | `input_conversion`  | function |
+| (I8)  | `body`              | function |
+| (I?)  | `output_conversion` | function |
+
+#### Constraints
+
+* (C?) `element_type(inputs...) = element_type(init_values...)`.
+* (C?) `baseline_element_type(inputs...) = baseline_element_type(results...)`.
+* (C?) `input_conversion` has type `tensor<E0>, ..., tensor<EN-1> ->
+       (tensor<E'0>, ..., tensor<E'N-1>)` where `Ei = element_type(inputs[i])`.
+* (C?) `body` has type `tensor<E0>, ..., tensor<EN-1>, tensor<E0>, ...,`
+       `tensor<EN-1>) -> (tensor<E0>, ..., tensor<EN-1>)` where
+       `Ei = element_type(output_types(input_conversion)[i])`.
+* (C?) `output_conversion` has type `tensor<E0>, ..., tensor<EN-1> ->
+       (tensor<E'0>, ..., tensor<E'N-1>)` where
+       `E'i = element_type(results[i])`.
+* (C?) `element_type(output_types(input_conversion)...) =
+       element_type(input_types(output_conversion)...)`.
+```
+
+### select_and_scatter
+
+This op originally takes two function arguments `select` and `scatter`. As the
+`select` function is supposed to perform a non-accumulative operation, we may
+not need additional conversion functions associated with `select`. But the
+`scatter` function needs be accompanied with `input_conversion` and
+`output_conversion` functions.
+
+```python
```

Comment:
Wrong line?

Reply:
done.

Reply:
I don't think this comment is addressed.  Why are you wrapping the following sections with python code block?